File: windows_phone/MpOS API/MpOS/Net/Endpoint/EndpointController.cs
/*******************************************************************************
 * Copyright (C) 2014 Philipp B. Costa
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *******************************************************************************/
using System.Net;
using Ufc.MpOS.Net.Endpoint.Service;
using Ufc.MpOS.Net.Rpc.Deploy;
using Ufc.MpOS.Net.Rpc.Model;
using Ufc.MpOS.Offload;
using Ufc.MpOS.Util.Device;

namespace Ufc.MpOS.Net.Endpoint
{
	public sealed class EndpointController
	{
		private readonly object mutex = new object();

		public static readonly int REPEAT_DISCOVERY_TASK = 20 * 1000;
		public static readonly int REPEAT_DECISION_MAKER = 35 * 1000;

		public static RpcProfile rpcProfile = new RpcProfile();

		public EndpointController(string secondaryEndpoint, bool decisionMakerActive, bool discoveryCloudlet)
		{
			this.decisionMakerActive = decisionMakerActive;
			
			SecondaryServer = new ServerContent(EndpointType.SECONDARY_SERVER);
			CloudletServer = new ServerContent(EndpointType.CLOUDLET);
			remoteAdvantage = false;

			if (secondaryEndpoint != null)
			{
				SecondaryServerIp(secondaryEndpoint);
				DiscoveryServiceSecondary();
			}
			if (discoveryCloudlet)
			{
				DiscoveryCloudletMulticast();
			}
		}

		public ServerContent SecondaryServer { get; private set; }
		public ServerContent CloudletServer { get; private set; }

		private DiscoveryCloudletMulticast discoveryCloudletMulticast = null;
		private DiscoveryService discoveryCloudletServer = null;
		private DiscoveryService discoveryInternetServer = null;

		private DynamicDecisionSystem dynamicDecisionSystem = null;
		private bool decisionMakerActive;

		private bool remoteAdvantage;
		public bool RemoteAdvantage
		{
			get
			{
				lock (mutex)
				{
					return remoteAdvantage;
				}
			}
			set
			{
				lock (mutex)
				{
					remoteAdvantage = value;
				}
			}
		}

		private void DiscoveryServiceSecondary()
		{
			if (discoveryInternetServer == null)
			{
				SecondaryServer.Clean();
				discoveryInternetServer = new DiscoveryService(SecondaryServer);
				discoveryInternetServer.Start();
			}
		}

		public void SecondaryServerIp(string ip)
		{
			SecondaryServer.Ip = IPAddress.Parse(ip);
		}

		public ServerContent CheckSecondaryServer()
		{
			if (MposFramework.Instance.DeviceController.IsOnline() && SecondaryServer.IsReady())
			{
				return SecondaryServer;
			}
			return null;
		}

		private void DiscoveryCloudletMulticast()
		{
			if (discoveryCloudletMulticast == null)
			{
				CloudletServer.Clean();
				discoveryCloudletMulticast = new DiscoveryCloudletMulticast();
				discoveryCloudletMulticast.Start();
			}
		}

		public void FoundCloudlet(string cloudletIp)
		{
			System.Diagnostics.Debug.WriteLine("#> [EndpointController]: Cloudlet address: " + cloudletIp);
			CloudletServer.Ip = IPAddress.Parse(cloudletIp);
			DiscoveryServiceCloudlet();
			discoveryCloudletMulticast = null;
		}

		private void DiscoveryServiceCloudlet()
		{
			if (discoveryCloudletServer == null)
			{
				discoveryCloudletServer = new DiscoveryService(CloudletServer);
				discoveryCloudletServer.Start();
			}
		}

		private ServerContent CheckCloudletServer()
		{
			if (MposFramework.Instance.DeviceController.ConnectionStatus(ConnectivityManager.TYPE_WIFI) && CloudletServer.IsReady())
			{
				return CloudletServer;
			}
			return null;
		}

		public void RediscoveryServices(ServerContent server)
		{
			if (server.Type == EndpointType.SECONDARY_SERVER)
			{
				DiscoveryServiceSecondary();
			}
			else if (server.Type == EndpointType.CLOUDLET)
			{
				DiscoveryCloudletMulticast();
			}
		}

		public ServerContent SelectPriorityServer(bool cloudletPriority)
		{
			if (cloudletPriority)
			{
				ServerContent server = CheckCloudletServer();
				if (server != null)
				{
					return server;
				}
				server = CheckSecondaryServer();
				if (server != null)
				{
					return server;
				}
			}
			else
			{
				ServerContent server = CheckSecondaryServer();
				if (server != null)
				{
					return server;
				}
				server = CheckCloudletServer();
				if (server != null)
				{
					return server;
				}
			}
			return null;
		}

		public void DeployService(ServerContent server)
		{
			new DeployService(server).Start();
		}

		public void StartDecisionMaker(ServerContent server){
			lock (mutex)
			{
				if (dynamicDecisionSystem == null && decisionMakerActive)
				{
					dynamicDecisionSystem = new DynamicDecisionSystem(server);
					dynamicDecisionSystem.Start();
				}
			}
		}

		//on UML = updateDdsEndpoint 
		public void UpdateDynamicDecisionSystemEndpoint(ServerContent server)
		{
			dynamicDecisionSystem.Server = server;
		}
	}
}

File: android/MpOS API/src/br/ufc/mdcc/mpos/net/endpoint/EndpointController.java
/*******************************************************************************
 * Copyright (C) 2014 Philipp B. Costa
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *******************************************************************************/
package br.ufc.mdcc.mpos.net.endpoint;

import java.util.Timer;

import android.content.Context;
import android.net.ConnectivityManager;
import android.util.Log;
import br.ufc.mdcc.mpos.MposFramework;
import br.ufc.mdcc.mpos.net.endpoint.service.DiscoveryCloudletMulticast;
import br.ufc.mdcc.mpos.net.endpoint.service.DiscoveryService;
import br.ufc.mdcc.mpos.net.exceptions.NetworkException;
import br.ufc.mdcc.mpos.net.rpc.deploy.DeployService;
import br.ufc.mdcc.mpos.net.rpc.model.RpcProfile;
import br.ufc.mdcc.mpos.offload.DynamicDecisionSystem;
import br.ufc.mdcc.mpos.util.Util;

/**
 * This class control many cycle-life services: 
 * - Discovery Service in local or remotable servers on internet or intranet.
 * - Deploy mobile app dependences in remote service.
 * - Decision Maker select.
 * - Etc.   
 * 
 * @author Philipp B. Costa
 */
public final class EndpointController {
	private final String clsName = EndpointController.class.getName();

	public static final int REPEAT_DISCOVERY_TASK = 20 * 1000;
	public static final int REPEAT_DECISION_MAKER = 35 * 1000;

	private Context context;

	private ServerContent secondaryServer;
	private ServerContent cloudletServer;

	// used for rediscovery services before discovery and using
	// help on mobility user!
	private DiscoveryService discoverySecondaryServer = null;
	private DiscoveryService discoveryCloudletServer = null;
	private DiscoveryCloudletMulticast discoveryCloudletMulticast = null;

	private Timer decisionMakerTimer;
	private DynamicDecisionSystem dynamicDecisionSystem;
	private boolean decisionMakerActive;
	private boolean remoteAdvantageExecution;

	public RpcProfile rpcProfile = new RpcProfile();

	public EndpointController(Context context, String internetIp, boolean decisionMakerActive, boolean discoveryCloudlet) throws NetworkException {
	    this.context = context;
        this.decisionMakerActive = decisionMakerActive;
	    
	    secondaryServer = new ServerContent(EndpointType.SECONDARY_SERVER);
		cloudletServer = new ServerContent(EndpointType.CLOUDLET);
		remoteAdvantageExecution = false;
        
        if(internetIp != null){
            setSecondaryServerIp(internetIp);
            discoveryServiceSecondary();
        }
        if(discoveryCloudlet){
            discoveryCloudletMulticast();
        }
	}

	private void discoveryServiceSecondary() {
		if(discoverySecondaryServer == null){
			secondaryServer.clean();
			discoverySecondaryServer = new DiscoveryService(secondaryServer);
			discoverySecondaryServer.setName("Discovery Internet Server");
			discoverySecondaryServer.start();
		}
	}

	private void setSecondaryServerIp(String ip) throws NetworkException {
		if (!Util.validateIpAddress(ip)) {
			throw new NetworkException("Invalid IP Address");
		}
		secondaryServer.setIp(ip);
	}

	public ServerContent getSecondaryServer() {
		return secondaryServer;
	}

	public ServerContent checkSecondaryServer() {
		if (MposFramework.getInstance().getDeviceController().isOnline() && secondaryServer.isReady()) {
			return secondaryServer;
		}
		return null;
	}

	private void discoveryCloudletMulticast() {
		if (context != null) {
			if (discoveryCloudletMulticast == null) {
			    cloudletServer.clean();
				discoveryCloudletMulticast = new DiscoveryCloudletMulticast(context);
				discoveryCloudletMulticast.setName("Discovery Cloudlet Multicast");
				discoveryCloudletMulticast.start();
			}
		}
	}

	public void foundCloudlet(String cloudletIp) {
		Log.i(clsName, "Cloudlet address: " + cloudletIp);

		cloudletServer.setIp(cloudletIp);
		discoveryServiceCloudlet();
        shutdownRediscoveryCloudletMulticast();
	}

	private void discoveryServiceCloudlet() {
		if (discoveryCloudletServer == null) {
			discoveryCloudletServer = new DiscoveryService(cloudletServer);
			discoveryCloudletServer.setName("Discovery Cloudlet Server");
			discoveryCloudletServer.start();
		}
	}

	public ServerContent getCloudletServer() {
		return cloudletServer;
	}

	private ServerContent checkCloudletServer() {
		if (MposFramework.getInstance().getDeviceController().connectionStatus(ConnectivityManager.TYPE_WIFI) && cloudletServer.isReady()) {
			return cloudletServer;
		}
		return null;
	}

	public void rediscoveryServices(ServerContent server) {
		if (server.getType() == EndpointType.SECONDARY_SERVER) {
			discoveryServiceSecondary();
		} else if (server.getType() == EndpointType.CLOUDLET) {
			discoveryCloudletMulticast();
		}
	}

	public ServerContent selectPriorityServer(boolean cloudletPriority) {
		if (cloudletPriority) {
			ServerContent server = checkCloudletServer();
			if (server != null) {
				return server;
			}
			server = checkSecondaryServer();
			if (server != null) {
				return server;
			}
		} else {
			ServerContent server = checkSecondaryServer();
			if (server != null) {
				return server;
			}
			server = checkCloudletServer();
			if (server != null) {
				return server;
			}
		}
		return null;
	}

	public void deployService(ServerContent server) {
		new DeployService(context, server).start();
	}

	public void shutdownRediscoveryInternetServer() {
		if(discoverySecondaryServer != null){
			discoverySecondaryServer.stopTask();
			discoverySecondaryServer.interrupt();
			discoverySecondaryServer = null;
		}
	}

	public void shutdownRediscoveryCloudletServer() {
		if (discoveryCloudletServer != null) {
			discoveryCloudletServer.stopTask();
			discoveryCloudletServer.interrupt();
			discoveryCloudletServer = null;
		}
	}

	private void shutdownRediscoveryCloudletMulticast() {
		if (discoveryCloudletMulticast != null) {
			discoveryCloudletMulticast.interrupt();
			discoveryCloudletMulticast = null;
		}
	}

	private void shutdownDecisionMaker() {
		if (decisionMakerTimer != null) {
			decisionMakerTimer.cancel();
			decisionMakerTimer.purge();
			decisionMakerTimer = null;
		}
	}

	//TODO: transf decision maker in thread...
	public synchronized void startDecisionMaker(ServerContent server) {
		if (decisionMakerTimer == null && decisionMakerActive) {
			decisionMakerTimer = new Timer("Decision Maker Watch");
			dynamicDecisionSystem = new DynamicDecisionSystem(context, server);
			decisionMakerTimer.schedule(dynamicDecisionSystem, 0, REPEAT_DECISION_MAKER);
		}
	}

	public synchronized void setRemoteAdvantageExecution(boolean remoteAdvantageExecution) {
		this.remoteAdvantageExecution = remoteAdvantageExecution;
	}

	public synchronized boolean isRemoteAdvantage() {
		return remoteAdvantageExecution;
	}

	//on UML -> updateDdsEndpoint 
	public void updateDynamicDecisionSystemEndpoint(ServerContent server) {
		dynamicDecisionSystem.setServer(server);
	}

	public void destroy() {
		shutdownRediscoveryCloudletMulticast();
		shutdownRediscoveryCloudletServer();
		shutdownRediscoveryInternetServer();
		shutdownDecisionMaker();

		secondaryServer.setIp(null);
		secondaryServer.clean();
		cloudletServer.setIp(null);
		cloudletServer.clean();
	}
}